superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx

```diff
@@ -66,47 +77,230 @@ describe('VisualizeModal', () => {
+    const sampleQuery = queries[0];
 
-    wrapper.setState({
-      chartType: mockChartTypeBarChart,
-      columns: mockColumns,
-      datasourceName: 'mockDatasourceName',
+    it('should require valid props parameters', () => {
+      const spy = sinon.spy(wrapper.instance(), 'setState');
+      wrapper.instance().setStateFromProps();
+      expect(spy.callCount).to.equal(0);
+      spy.restore();
+    });
+    it('should set columns state', () => {
+      wrapper.instance().setStateFromProps({ query: sampleQuery });
+      expect(wrapper.state().columns).to.deep.equal(mockColumns);
+    });
+  });
+
+  describe('datasourceName', () => {
+    const wrapper = getVisualizeModalWrapper();
+    let stub;
+    beforeEach(() => {
+      stub = sinon.stub(shortid, 'generate').returns('abcd');
+    });
+    afterEach(() => {
+      stub.restore();
+    });
+
+    it('should generate data source name from query', () => {
+      const sampleQuery = queries[0];
+      const name = wrapper.instance().datasourceName();
+      expect(name).to.equal(`${sampleQuery.user}-${sampleQuery.db}-${sampleQuery.tab}-abcd`);
+    });
+    it('should generate data source name with empty query', () => {
+      wrapper.setProps({ query: {} });
+      const name = wrapper.instance().datasourceName();
+      expect(name).to.equal('undefined-abcd');
+    });
+  });
+
+  describe('mergedColumns', () => {
+    const wrapper = getVisualizeModalWrapper();
+    const oldColumns = {
+      ds: 1,
+      gender: 2,
+    };
+
+    it('should merge by column name', () => {
+      wrapper.setState({ columns: {} });
+      const mc = wrapper.instance().mergedColumns();
+      expect(mc).to.deep.equal(mockColumns);
+    });
+    it('should not override current state', () => {
+      wrapper.setState({ columns: oldColumns });
+
+      const mc = wrapper.instance().mergedColumns();
+      expect(mc.ds).to.equal(oldColumns.ds);
+      expect(mc.gender).to.equal(oldColumns.gender);
+    });
+  });
+
+  describe('validate', () => {
+    const wrapper = getVisualizeModalWrapper();
+    let columnsStub;
+    beforeEach(() => {
+      columnsStub = sinon.stub(wrapper.instance(), 'mergedColumns');
+    });
+    afterEach(() => {
+      columnsStub.restore();
+    });
+
+    it('should validate column name', () => {
+      columnsStub.returns(mockColumns);
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(0);
+      wrapper.instance().mergedColumns.restore();
+    });
+    it('should hint invalid column name', () => {
+      columnsStub.returns({
+        '&': 1,
+      });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
+      wrapper.instance().mergedColumns.restore();
     });
+    it('should hint empty chartType', () => {
+      columnsStub.returns(mockColumns);
+      wrapper.setState({ chartType: null });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
```

Comment:
could we be more specific and check that the hints array contains the correct hint...